Documentation update (#14)
jenkinsci/datadog-plugin: README.md

# Jenkins Datadog Plugin

[![Build Status](https://dev.azure.com/datadoghq/jenkins-datadog-plugin/_apis/build/status/DataDog.jenkins-datadog-plugin?branchName=master)](https://dev.azure.com/datadoghq/jenkins-datadog-plugin/_build/latest?definitionId=18&branchName=master)

> A Jenkins plugin used to forward automatically metrics, events, and service checks to a Datadog account.

Reviewer: ```suggestion
A Jenkins plugin for automatically forwarding metrics, events, and service checks to a Datadog account.
```

> **Note**: There is a [Jenkins CI Plugin page][1] for this plugin, but it refers to this documentation.

Reviewer: ```suggestion
**Note**: The [Jenkins CI plugin page][1] for this plugin references this documentation.
```

## Setup

### Installation

_This plugin requires [Jenkins 1.580.1][2] or newer._

This plugin can be installed from the [Update Center][3] (found at `Manage Jenkins -> Manage Plugins`) in your Jenkins installation:

> 1. Select the `Available` tab, search for `Datadog` and look for `Datadog Plugin`.

Reviewer: ```suggestion
1. Select the `Available` tab, search for `Datadog`, and select the checkbox next to `Datadog Plugin`.
```

> 2. Check the checkbox next to it, and install via your preference by using one of the two install buttons at the bottom of the screen.

Reviewer: ```suggestion
2. Install the plugin by using one of the two install buttons at the bottom of the screen.
```

> 3. Check that the plugin has been successfully installed by searching for `Datadog Plugin` on the `Installed` tab. If the plugin has been successfully installed, continue on to the configuration step, described below.

Reviewer: ```suggestion
3. To verify the plugin is installed, search for `Datadog Plugin` on the `Installed` tab. After the plugin is installed successfully, continue to the configuration section below.
```

> **Note**: If you do not see the version of `Datadog Plugin` that you are expecting, make sure you have run `Check Now` from the `Manage Jenkins -> Manage Plugins` screen.

Reviewer: ```suggestion
**Note**: If you see an unexpected version of the `Datadog Plugin`, run `Check Now` from the `Manage Jenkins -> Manage Plugins` screen.
```

### Configuration

You can use two ways to configure your plugin to submit data to Datadog:

* Sending the data directly to Datadog through HTTP.
* Using a DogStatsD server that acts as a forwarder between Jenkins and Datadog.

> This configuration can be done from the [plugin user interface](#plugin-user-interface), thanks to a [Groovy script](#groovy-script), or through [Environment variables](#environment-variables).

Reviewer: ```suggestion
The configuration can be done from the [plugin user interface](#plugin-user-interface) with a [Groovy script](#groovy-script), or through [environment variables](#environment-variables).
```

#### Plugin user interface

To configure your Datadog Plugin, navigate to the `Manage Jenkins -> Configure System` page on your Jenkins installation. Once there, scroll down to find the `Datadog Plugin` section:

##### HTTP forwarding {#http-forwarding-plugin}

> 1. Click the "Use Datadog API URL and Key to report to Datadog" radio button (selected by default)

Reviewer: ```suggestion
1. Select the radio button next to **Use Datadog API URL and Key to report to Datadog** (selected by default).
```

> 2. Find your API Key from the [API Keys][4] page on your Datadog account, and copy/paste it into the `API Key` textbox on the Jenkins configuration screen.

Reviewer: ```suggestion
2. Use your [Datadog API key][4] in the `API Key` textbox on the Jenkins configuration screen.
```

> 3. You can test that your API Key works by pressing the `Test Key` button, on the Jenkins configuration screen, directly below the API Key textbox.

Reviewer: ```suggestion
3. Test your Datadog API key by using the `Test Key` button on the Jenkins configuration screen directly below the API key textbox.
```

4. Save your configuration.

##### DogStatsD forwarding {#dogstatsd-forwarding-plugin}

> 1. Click the **Use a DogStatsD Server to report to Datadog** radio button.

Reviewer: ```suggestion
1. Select the radio button next to **Use a DogStatsD Server to report to Datadog**.
```

> 2. Specify both your DogStatD server `hostname` and `port`

Reviewer: ```suggestion
2. Specify your DogStatD server `hostname` and `port`.
```

3. Save your configuration.

#### Groovy script

> Configure your Datadog plugin to forward its data through HTTP or DogStatsD using the Groovy scripts below. Configuring the plugin this way might be useful if you're running your Jenkins Master in a Docker container using the [Official Jenkins Docker Image][5] or any derivative that supports plugins.txt and Groovy init scripts.

Reviewer: ```suggestion
Configure your Datadog plugin to forward data through HTTP or DogStatsD using the Groovy scripts below. Configuring the plugin this way might be useful if you're running your Jenkins Master in a Docker container using the [official Jenkins Docker image][5] or any derivative that supports `plugins.txt` and Groovy init scripts.
```

##### HTTP forwarding {#http-forwarding-groovy-script}

```groovy
import jenkins.model.*
import org.datadog.jenkins.plugins.datadog.DatadogGlobalConfiguration

def j = Jenkins.getInstance()
def d = j.getDescriptor("org.datadog.jenkins.plugins.datadog.DatadogGlobalConfiguration")

// If you want to use Datadog API URL and Key to report to Datadog
d.setReportWith('HTTP')
d.setTargetApiURL('https://api.datadoghq.com/api/')
d.setTargetApiKey('<DATADOG_API_KEY>')

// Customization, see dedicated section below
d.setBlacklist('job1,job2')

// Save config
d.save()
```

##### DogStatsD forwarding {#dogstatsd-forwarding-groovy-script}

```groovy
import jenkins.model.*
import org.datadog.jenkins.plugins.datadog.DatadogGlobalConfiguration

def j = Jenkins.getInstance()
def d = j.getDescriptor("org.datadog.jenkins.plugins.datadog.DatadogGlobalConfiguration")

d.setReportWith('DSD')
d.setTargetHost('localhost')
d.setTargetPort(8125)

// Customization, see dedicated section below
d.setBlacklist('job1,job2')

// Save config
d.save()
```

#### Environment variables

> Configure your Datadog plugin using environment variables by specifying the `DATADOG_JENKINS_PLUGIN_REPORT_WITH` variable which specifies which report mechanism you want to use.

Reviewer: ```suggestion
Configure your Datadog plugin using environment variables with the `DATADOG_JENKINS_PLUGIN_REPORT_WITH` variable, which specifies the report mechanism to use.
```

> ##### HTTP Forwarding {#http-forwarding-env}

Reviewer: ```suggestion
##### HTTP forwarding {#http-forwarding-env}
```

1. Set the `DATADOG_JENKINS_PLUGIN_REPORT_WITH` variable to `HTTP`.

> 2. Set the `DATADOG_JENKINS_PLUGIN_TARGET_API_URL` parameter which specifies the Datadog API Endpoint to report to. Default value is `https://api.datadoghq.com/api/`.

Reviewer: ```suggestion
2. Set the `DATADOG_JENKINS_PLUGIN_TARGET_API_URL` variable, which specifies the Datadog API endpoint (defaults `https://api.datadoghq.com/api/`).
```

> 3. Set the `DATADOG_JENKINS_PLUGIN_TARGET_API_KEY` parameter which specifies your Datadog API key in order to report to your Datadog account. Get your API Key from the [Datadog API Keys page][4].

Reviewer: ```suggestion
3. Set the `DATADOG_JENKINS_PLUGIN_TARGET_API_KEY` variable, which specifies your [Datadog API key][4].
```

##### DogStatsD forwarding {#dogstatsd-forwarding-env}

1. Set the `DATADOG_JENKINS_PLUGIN_REPORT_WITH` variable to `DSD`.

> 2. Set the `DATADOG_JENKINS_PLUGIN_TARGET_HOST` variable which specifies the DogStatsD Server host to report to. Default value is `localhost`.

Reviewer: ```suggestion
2. Set the `DATADOG_JENKINS_PLUGIN_TARGET_HOST` variable, which specifies the DogStatsD server host (defaults to `localhost`).
```

> 3. Set the `DATADOG_JENKINS_PLUGIN_TARGET_PORT` variable which specifies the DogStatsD Server port to report to. Default value is `8125`.

Reviewer: ```suggestion
3. Set the `DATADOG_JENKINS_PLUGIN_TARGET_PORT` variable, which specifies the DogStatsD server port (defaults to `8125`).
```

#### Logging

> Logging is done by utilizing the `java.util.Logger`, which follows the [best logging practices for Jenkins][6]. In order to obtain logs, follow the directions listed in [the Jenkins logging documentation][6]. When adding a Logger, all Datadog plugin functions start with `org.datadog.jenkins.plugins.datadog.` and the function name you are after should autopopulate. As of this writing, the only function available was `org.datadog.jenkins.plugins.datadog.listeners.DatadogBuildListener`.

Reviewer: ```suggestion
Logging is done by utilizing the `java.util.Logger`, which follows the [best logging practices for Jenkins][6]. To obtain logs, follow the directions in the [Jenkins logging documentation][6]. When adding a logger, all Datadog plugin functions start with `org.datadog.jenkins.plugins.datadog.` and the function name you are after should autopopulate. As of this writing, the only function available was `org.datadog.jenkins.plugins.datadog.listeners.DatadogBuildListener`.
```

## Customization

> ### Global Customization

Reviewer: ```suggestion
### Global customization
```

### Events

#### Default Events Type

| Event Name      | Triggered on              | Default Tags                                                               | Associated RATE metric  |
|-----------------|---------------------------|----------------------------------------------------------------------------|-------------------------|
| Build Started   | `RunListener#onStarted`   | `job`, `node`, `branch`                                                    | `jenkins.job.started`   |
| Build Aborted   | `RunListener#onDeleted`   | `job`, `node`, `branch`                                                    | `jenkins.job.aborted`   |
| Build Completed | `RunListener#onCompleted` | `job`, `node`, `branch`, `result` (Git Branch, SVN revision or CVS branch) | `jenkins.job.completed` |

#### Source Control Management Events Type

| Event Name   | Triggered on             | Default Tags            | Associated RATE metric |
|--------------|--------------------------|-------------------------|------------------------|
| SCM Checkout | `SCMListener#onCheckout` | `job`, `node`, `branch` | `jenkins.scm.checkout` |

#### Systems Events Type

> | Event Name                   | Triggered on                            | Associated RATE metric                 |

Author: ```suggestion
| Event name                   | Triggered on                            | Associated RATE metric                 |
```

| Computer Offline             | `ComputerListener#onOffline`            | `jenkins.computer.online`              |
| Computer TemporarilyOnline   | `ComputerListener#onTemporarilyOnline`  | `jenkins.computer.temporarily_online`  |
| Computer TemporarilyOffline  | `ComputerListener#onTemporarilyOffline` | `jenkins.computer.temporarily_offline` |
| Computer LaunchFailure       | `ComputerListener#onLaunchFailure`      | `jenkins.computer.launch_failure`      |
| Item Created                 | `ItemListener#onCreated`                | `jenkins.item.created`                 |
| Item Deleted                 | `ItemListener#onDeleted`                | `jenkins.item.deleted`                 |
| Item Updated                 | `ItemListener#onUpdated`                | `jenkins.item.updated`                 |
| Item Copied                  | `ItemListener#onCopied`                 | `jenkins.item.copied`                  |
| ItemListener LocationChanged | `ItemListener#onLocationChanged`        | `jenkins.item.location_changed`        |
| Config Changed               | `SaveableListener#onChange`             | `jenkins.config.changed`               |

#### Security Events Type

> | Event Name                  | Triggered on                            | Associated RATE metric       |

Author: ```suggestion
| Event name                  | Triggered on                            | Associated RATE metric       |
```

## Release Process

### Overview

The [DataDog/jenkins-datadog-plugin][8] repository handles the most up-to-date changes made to the Datadog Plugin, as well as issue tickets revolving around that work. Releases are merged to the [Jenkins-CI git repo for our plugin][9], and represents the source used for plugin releases found in the [Update Center][3] in your Jenkins installation.

Every commit to the [DataDog/jenkins-datadog-plugin][8] repository triggers a Jenkins build on our internal Jenkins installation.

A list of releases is available at [jenkinsci/datadog-plugin/releases][10].

### How to Release

To release a new plugin version, change the project version in the [pom.xml][11] from `x.x.x-SNAPSHOT` to the updated version number you would like to see. Add an entry for the new release number to the [CHANGELOG.md][12] file, and ensure that all the changes are listed accurately. Then run the `jenkins-datadog-plugin-release` job in the Jenkins installation. If the job completes successfully, then the newly updated plugin should be available from the Jenkins [Update Center][3] within ~4 hours (plus mirror propogation time).

## Issue Tracking

> The Github's built in issue tracking system is used to track all issues tickets relating to this plugin, found [DataDog/jenkins-datadog-plugin/issues][13]. However, given how Jenkins Plugins are hosted, there may be issues that are posted to JIRA as well. You can check [this jenkins issue][14] for those issue postings.

Reviewer: ```suggestion
Github's built-in issue tracking system is used to track all issues relating to this plugin: [DataDog/jenkins-datadog-plugin/issues][13]. However, given how Jenkins plugins are hosted, there may be issues that are posted to JIRA as well. You can check [this jenkins issue][14] for those issue postings.
```

**Note**: [Unresolved issues on JIRA mentioning Datadog.][15].

## Changes

See the [CHANGELOG.md][12].

## How to contribute code

First of all and most importantly, **thank you** for sharing.

If you want to submit code, fork this repository and submit pull requests against the `master` branch. For more information, checkout the [contributing guidelines][16] for the Datadog Agent.

Check out the [development document][17] for tips on spinning up a quick development environment locally.

> ## Manual Testing

Reviewer: ```suggestion
## Manual testing
```

> In order to keep track of some testing procedures for ensuring proper functionality of the Datadog Plugin on Jenkins, there is a [testing document][17].

Reviewer: ```suggestion
To keep track of testing procedures for ensuring proper functionality of the Datadog Plugin on Jenkins, there is a [testing document][17].
```
